### ISSUE TITLE
hg_prompt is slow for large repos

### ISSUE THREAD

**Issue by satayev** [dimo414/prompt.gem#14]
It boils down to two invocations:  `hg branch` and `hg stat`. Both can be significantly slow for large repos which makes it difficult to use prompt.gem there.

**Comment by satayev** [comment]
See https://github.com/dimo414/prompt.gem/pull/15 for a fix for branch invocation.

For stat invocation, I am not sure there is an easy fix. I personally just ended up removing it from my local prompt.gem, perhaps it's worth adding a base.conf.sh variable to conditionally execute it.

**Comment by dimo414** [comment]
Thanks for filing, yes `hg_prompt` is painfully slow. I've been meaning to revisit it but I have fewer Mercurial repos these days so it hasn't been as noticeable for me.

Also (as you may be aware) these expensive functions are cached by http://github.com/dimo414/bash-cache, so the slowness shouldn't be as noticeable in a shell that's actively being used. Tuning the [cache duration](https://github.com/dimo414/prompt.gem/blob/a1b0c9a/env_functions.sh#L47) might be an easy improvement, in addition to investigating better implementations.

> I personally just ended up removing it from my local prompt.gem

FWIW the intended way to do so is to remove it from the `ENV_INFO` var in your `local.conf.sh`. Removing it from your local checkout is fine too, obviously, but you may have to resolve merge conflicts at some point.

**Comment by satayev** [comment]
Agreed, bash-cache is amazing! I am just pushing some general changes I made to profilegem, and thought some of these changes are generic / good enough to be shared.

I do keep hg_prompt in ENV_INFO; however I removed the call to `hg stat` and added branch & bookmark calls. So, I know what feature I am working on, but I don't know if the working dir is dirty or not. (Instead of defining my own hg_prompt separately.)

P.S. speaking of hg_prompt, it may be worth including `hg bookmark` information as well; but that's a separate issue.

**Comment by dimo414** [comment]
Gotcha, yes customizing the implementation of `hg_prompt` makes sense. And feel free to send a PR adding `hg bookmark` :)
